fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java

```diff
@@ -3505,130 +3497,263 @@ public void replayRecoverPartition(RecoverInfo info) {
-            int totalTaskNum = index.getTablets().size() * replicationNum;
-            MarkedCountDownLatch<Long, Long> countDownLatch = new MarkedCountDownLatch<Long, Long>(totalTaskNum);
-            AgentBatchTask batchTask = new AgentBatchTask();
-            for (Tablet tablet : index.getTablets()) {
-                long tabletId = tablet.getId();
-                for (Replica replica : tablet.getReplicas()) {
-                    long backendId = replica.getBackendId();
-                    countDownLatch.addMark(backendId, tabletId);
-                    CreateReplicaTask task = new CreateReplicaTask(backendId, dbId, tableId,
-                            partitionId, indexId, tabletId,
-                            shortKeyColumnCount, schemaHash,
-                            version, versionHash,
-                            keysType,
-                            storageType, storageMedium,
-                            schema, bfColumns, bfFpp,
-                            countDownLatch,
-                            indexes,
-                            isInMemory,
-                            tabletType);
-                    task.setStorageFormat(storageFormat);
-                    batchTask.addTask(task);
-                    // add to AgentTaskQueue for handling finish report.
-                    // not for resending task
-                    AgentTaskQueue.addTask(task);
-                }
+            TabletMeta tabletMeta = new TabletMeta(db.getId(), table.getId(), partitionId, indexId, indexMeta.getSchemaHash(),
+                                                   storageMedium);
+            createTablets(db.getClusterName(), index, ReplicaState.NORMAL, distributionInfo, partition.getVisibleVersion(),
+                    partition.getVisibleVersionHash(), replicationNum, tabletMeta, tabletIdSet);
+            if (index.getId() != table.getBaseIndexId()) {
+                // add rollup index to partition
+                partition.createRollupIndex(index);
             }
-            AgentTaskExecutor.submit(batchTask);
+        }
+        return partition;
+    }
 
-            // estimate timeout
-            long timeout = Config.tablet_create_timeout_second * 1000L * totalTaskNum;
-            timeout = Math.min(timeout, Config.max_create_table_timeout_second * 1000);
+    private void buildPartitions(Database db, OlapTable table, List<Partition> partitions) throws DdlException {
+        if (partitions.isEmpty()) {
+            return;
+        }
+        int numAliveBackends = systemInfo.getBackendIds(true).size();
+        int numReplicas = 0;
+        for (Partition partition : partitions) {
+            numReplicas += partition.getReplicaCount();
+        }
+
+        if (partitions.size() >= 3 && numAliveBackends >= 3 && numReplicas >= numAliveBackends * 500) {
+            LOG.info("creating {} partitions of table {} concurrently", partitions.size(), table.getName());
```

Comment:
have you test the number 500? Can one backend create 500 tablets in max_create_table_timeout_second (60)?

Reply:
Should this number be configurable?

Reply:
> have you test the number 500? Can one backend create 500 tablets in max_create_table_timeout_second (60)?

`max_create_table_timeout_second` is the maximum timeout in seconds applied for a SINGLE index, while `numAliveBackends` is the number of tasks among ALL indexes.

Reply:
> Should this number be configurable?

No such plan yet